python/tvm/relay/backend/contrib/tidl.py

```diff
@@ -0,0 +1,1403 @@
+                if self.rename_starting_from_0:
+                    subgraph_func = VarRenamer(new_name).visit(self.mod[name])
+                    subgraph_func = subgraph_func.with_attr("global_symbol", new_name)
+                    self.new_mod[subgraph_gv] = subgraph_func
+                else:
+                    self.new_mod[subgraph_gv] = self.mod[name]
+                return subgraph_gv(*args)
+        return super().visit_call(call)
+
+def PruneSubgraphsWithMoreThanOneInput(mod, compiler="tidl"):
+    subgraph_names_to_remove = []
+    # Remove subgraphs with more than 1 input or tuple inputs.
+    for subgraph in mod.get_global_vars():
+        name = subgraph.name_hint
+        if not mod[name].attrs or mod[name].attrs["Compiler"] != compiler:
+            continue
+        print("SUBGRAPH PARAMS", mod[name].params)
+        if len(mod[name].params) != 1 or isinstance(mod[name].params[0].checked_type, relay.TupleType):
+            subgraph_names_to_remove.append(name)
+    print("Removing subgraphs due to having more than one input:", subgraph_names_to_remove)
+    new_mod = tvm.IRModule()
+    new_mod["main"] = SubgraphRemover(subgraph_names_to_remove, mod, new_mod).visit(mod["main"])
+    return new_mod
+
+def PruneSubgraphs(mod, compiler="tidl", num_subgraphs_to_keep=4):
+    subgraph_with_macs = []
+    for subgraph in mod.get_global_vars():
+        name = subgraph.name_hint
+        if not mod[name].attrs or mod[name].attrs["Compiler"] != compiler:
+            continue
+        num_macs = relay.analysis.get_total_mac_number(mod[name])
+        subgraph_with_macs.append([name, num_macs])
+    subgraph_with_macs = sorted(subgraph_with_macs, key=lambda x: int(x[1]))
+    print("Subgraphs with computed # of MACS:", subgraph_with_macs)
+    subgraphs_to_remove = subgraph_with_macs[:-num_subgraphs_to_keep]
+    print("Will remove these subgraphs:", subgraphs_to_remove)
+    subgraph_names_to_remove = set([x[0] for x in subgraphs_to_remove])
+    # Create new pruned module
+    new_mod = tvm.IRModule()
+    new_mod["main"] = SubgraphRemover(subgraph_names_to_remove, mod, new_mod).visit(mod["main"])
+    return new_mod
+
+class TIDLCompiler:
+    """TIDL compiler module.
+
+    This module tries to compile a given Relay IR graph to deploy on devices with TIDL.
+    If compilation for TIDL succeeds, artifacts for heterogeneous compute with TIDL
+    will be generated.
+
+    Parameters
+    ----------
+    platform : string
+        The platform to deploy the graph on.
+    version : tuple
+        The Processor-SDK version for the platform.
+    **kwargs : keyword arguments to pass what's needed for Relay IR graph conversion
+        num_tidl_subgraphs : int
+            Number of subgraphs to run on TIDL
+        data_layout : string
+            Data layout, "NCHW" or "NHWC"
+        artifacts_folder : string
+            Folder to hold TIDL artifacts
+    """
+
+    def __init__(self, platform, version, **kwargs):

```

Comment:
Let's make the `kwargs` into regular args so we can have default values.

Reply:
We need kwargs because AM57x and future J7 devices will likely have different init arguments, and we want to keep enable API the same.

Reply:
@trevor-m , what about doing the following to have default values:
    def __init__(self, platform, version, **kwargs):
        if platform == "AM57" and version >= (6,3):
            self.num_tidl_subgraphs = kwargs.get("num_tidl_subgraphs", 1)
            self.data_layout = kwargs.get("data_layout", "NCHW")
            self.artifacts_folder = kwargs.get("artifacts_folder", None)
            self.tidl_tools_path = kwargs.get("tidl_tools_path", None)

Reply:
We also need to make sure a user provide the required args which are `artifacts_folder` and `tidl_tools_path`. Having those as standard positional args instead of kwargs naturally solves this. Otherwise we need to explicitly check they are provided.

We can do something like this to solve default issue and required args issue:

```
if platform == "AM57" and version >= (6,3):
  # Set default values for AM57 6.3
  self.num_tidl_subgraphs = 4
  self.data_layout = "NCHW"
  self.artifacts_folder = None
  self.tidl_tools_path = None
  for key in ('num_tidl_subgraphs', 'data_layout', 'artifacts_folder', 'tidl_tools_path'):
    if key in kwargs:
      setattr(self, key, kwargs[key])
  assert self.artifacts_folder, "artifacts_folder must be specified"
  assert self.tidl_tools_path, "tidl_tools_path must be specified"
```

Reply:
This is fine, but tidl_tools_path doesn't have to be provided as it is not available in CI.

Reply:
This is fine. We will not assert on tidl_tools_path which is not available in CI.